// @geckos.io/server/lib/httpServer/httpServer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var url_1 = __importDefault(require("url"));
var setCors_1 = __importDefault(require("./setCors"));
var parseBody_1 = __importDefault(require("./parseBody"));
var HttpServer = function (server, connectionsManager, cors) {
    var prefix = '.wrtc';
    var version = 'v1';
    var root = "/" + prefix + "/" + version;
    var rootRegEx = new RegExp("/" + prefix + "/" + version);
    var evs = server.listeners('request').slice(0);
    server.removeAllListeners('request');
    server.on('request', function (req, res) { return __awaiter(void 0, void 0, void 0, function () {
        var pathname, method, i, path1, path2, body, error_1, connection, id, iceConnectionState, peerConnection, remoteDescription, localDescription, signalingState, error_2, ids, id, connection, connectionJSON, error_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    pathname = req.url ? url_1.default.parse(req.url, true).pathname : undefined;
                    method = req.method;
                    // if the request is not part of the rootRegEx,
                    // trigger the other server's (Express) events.
                    if (!pathname || !rootRegEx.test(pathname)) {
                        for (i = 0; i < evs.length; i++) {
                            evs[i].call(server, req, res);
                        }
                    }
                    if (!(pathname && rootRegEx.test(pathname))) return [3 /*break*/, 15];
                    path1 = pathname === root + "/connections";
                    path2 = new RegExp(prefix + "/" + version + "/connections/[0-9a-zA-Z]+/remote-description").test(pathname);
                    setCors_1.default(req, res, cors);
                    if (req.method === 'OPTIONS') {
                        res.writeHead(200);
                        res.end();
                        return [2 /*return*/];
                    }
                    body = '';
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, parseBody_1.default(req)];
                case 2:
                    body = (_a.sent());
                    return [3 /*break*/, 4];
                case 3:
                    error_1 = _a.sent();
                    res.writeHead(400);
                    res.end();
                    return [2 /*return*/];
                case 4:
                    res.on('error', function (error) {
                        console.error(error.message);
                        res.writeHead(500);
                        res.end();
                        return;
                    });
                    res.setHeader('Content-Type', 'application/json');
                    if (!(pathname && method)) return [3 /*break*/, 15];
                    if (!(method === 'POST' && path1)) return [3 /*break*/, 9];
                    _a.label = 5;
                case 5:
                    _a.trys.push([5, 8, , 9]);
                    return [4 /*yield*/, connectionsManager.createConnection()
                        // create the offer
                    ];
                case 6:
                    connection = _a.sent();
                    // create the offer
                    return [4 /*yield*/, connection.doOffer()];
                case 7:
                    // create the offer
                    _a.sent();
                    id = connection.id, iceConnectionState = connection.iceConnectionState, peerConnection = connection.peerConnection, remoteDescription = connection.remoteDescription, localDescription = connection.localDescription, signalingState = connection.signalingState;
                    res.write(JSON.stringify({
                        id: id,
                        iceConnectionState: iceConnectionState,
                        peerConnection: peerConnection,
                        remoteDescription: remoteDescription,
                        localDescription: localDescription,
                        signalingState: signalingState
                    }));
                    res.end();
                    return [2 /*return*/];
                case 8:
                    error_2 = _a.sent();
                    console.error(error_2.message);
                    res.statusCode = 500;
                    res.end();
                    return [2 /*return*/];
                case 9:
                    if (!(method === 'POST' && path2)) return [3 /*break*/, 14];
                    ids = pathname.match(/[0-9a-zA-Z]{24}/g);
                    if (!(ids && ids.length === 1)) return [3 /*break*/, 13];
                    id = ids[0];
                    connection = connectionsManager.getConnection(id);
                    if (!connection) {
                        res.statusCode = 404;
                        res.end();
                        return [2 /*return*/];
                    }
                    _a.label = 10;
                case 10:
                    _a.trys.push([10, 12, , 13]);
                    return [4 /*yield*/, connection.applyAnswer(JSON.parse(body))];
                case 11:
                    _a.sent();
                    connectionJSON = connection.toJSON();
                    res.write(JSON.stringify(connectionJSON.remoteDescription));
                    res.end();
                    return [2 /*return*/];
                case 12:
                    error_3 = _a.sent();
                    console.error(error_3.message);
                    res.statusCode = 400;
                    res.end();
                    return [2 /*return*/];
                case 13: return [3 /*break*/, 15];
                case 14:
                    res.writeHead(404);
                    res.end();
                    return [2 /*return*/];
                case 15: return [2 /*return*/];
            }
        });
    }); });
};
exports.default = HttpServer;
